### ISSUE TITLE
My first unikernel - Thomas Leonard's blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2271]
My first unikernel - Thomas Leonard's blog<br>
<br>
http://bit.ly/1mXPp0q<br>
<br>
Thomas Leonard<br>
<br>

<div>
<div>
<p>I wanted to make a simple REST service for queuing file uploads, deployable as a virtual machine. The traditional way to do this is to download a Linux cloud image, install the software inside it, and deploy that. Instead I decided to try a <a href="http://queue.acm.org/detail.cfm?id=2566628">unikernel</a>.</p>

<p>Unikernels promise some interesting benefits. The Ubuntu 14.04 amd64-disk1.img cloud image is 243 MB unconfigured, while the unikernel ended up at just 5.2 MB (running the queue service). Ubuntu runs a large amount of C code in security-critical places, while the unikernel is almost entirely type-safe OCaml. And besides, trying new things is fun.</p>

<p>( this post also appeared on <a href="http://www.reddit.com/r/programming/comments/2c1soi/my_first_unikernel_created_in_ocaml_and_mirage/">Reddit</a> and <a href="https://news.ycombinator.com/item?id=8109485">Hacker News</a> )</p>

<p><strong>Table of Contents</strong></p>

<p>Regular readers will know that a few months ago I began a new job at Cambridge University.
Working for an author of Real World OCaml and leader of OCaml Labs, on a project building pure-OCaml distributed systems, who found me through my blog posts about learning OCaml, I thought they might want me to write some OCaml.</p>

<p>But no.
They’ve actually had me <a href="http://openmirage.org/blog/introducing-xen-minios-arm">porting the tiny Mini-OS kernel to ARM</a>, using a mixture of C and assembler, to let <a href="http://openmirage.org/">the Mirage unikernel</a> run on ARM devices.
Of course, I got curious and wanted to write a Mirage application for myself…</p>

<h2 id="introduction">Introduction</h2>

<p>Linux, like many popular operating systems, is a <em>multi-user</em> system.
This design dates back to the early days of computing, when a single expensive computer, running a single OS, would be shared between many users.
The goal of the kernel is to protect itself from its users, and to protect the users from each other.</p>

<p>Today, computers are cheap and many people own several.
Even when a physical computer is shared (e.g. in cloud computing), this is typically done by running multiple virtual machines, each serving a single user.
Here, protecting the OS from its (usually single) application is pointless.</p>

<p>Removing the security barrier between the kernel and the application greatly simplifies things;
we can run the whole system (kernel + application) as a single, privileged, executable - a <em>unikernel</em>.</p>

<p>And while we’re rewriting everything anyway, we might as well replace C with a modern memory safe language, eliminating whole classes of bugs and security vulnerabilities, allowing decent error reporting, and providing structured data types throughout.</p>

<p>In the past, two things have made writing a completely new OS impractical:</p>

<ul>
  <li>Legacy applications won’t run on it.</li>
  <li>It probably won’t support your hardware.</li>
</ul>

<p>Virtualisation removes both obstacles:
legacy applications can run in their own legacy VMs, and
drivers are only needed for the virtual devices - e.g.
a single network driver and a single block driver will cover all real network cards and hard drives.</p>

<h3 id="a-hello-world-kernel">A hello world kernel</h3>

<p>The <a href="http://openmirage.org/wiki/install">mirage tutorial</a> starts by showing the easy, fully-automated way to build a unikernel.
If you want to get started quickly you may prefer to read that and skip this section, but since one of the advantages of unikernels is their relative simplicity, let’s do things the “hard” way first to understand how it works behind the scenes.</p>

<p>Here’s the normal “hello world” program in OCaml:</p>

<div><figure><figcaption><span>hw.ml </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
</pre></td>
<td><pre><code><span><span>let</span> <span>()</span> <span>=</span>
</span><span>  <span>print_endline</span> <span>"Hello, world!"</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>To compile and run as a normal application, we’d do:</p>

<pre><code>$ ocamlopt hw.ml -o hw
$ ./hw 
Hello, world!
</code></pre>

<p>How can we make a unikernel that does the equivalent?
As it turns out, the above code works unmodified (though the Mirage people might frown at you for doing it this way).
We compile hw.ml to a <code>hw.native.o</code> file and then link with the unikernel libraries instead of the standard C library:</p>

<pre><code>$ export OPAM_DIR=$(opam config var prefix)
$ export PKG_CONFIG_PATH=$OPAM_DIR/lib/pkgconfig
$ ocamlopt -output-obj -o hw.native.o hw.ml
$ ld -d -static -nostdlib --start-group \
    $(pkg-config --static --libs openlibm libminios-xen) \
    hw.native.o \
    $OPAM_DIR/lib/mirage-xen/libocaml.a \
    $OPAM_DIR/lib/mirage-xen/libxencaml.a \
    --end-group \
    $(gcc -print-libgcc-file-name) \
    -o hw.xen
</code></pre>

<p>We now have a kernel image, <code>hw.xen</code>, which can be booted as a VM under the <a href="http://www.xenproject.org/developers/teams/hypervisor.html">Xen hypervisor</a> (as used by Amazon, Rackspace, etc to host VMs). But first, let’s look at the libraries we added:</p>

<dl>
  <dt>openlibm</dt>
  <dd>This is a standard maths library. It provides functions such as <code>sin</code>, <code>cos</code>, etc.</dd>
  <dt>libminios-xen</dt>
  <dd>This provides the architecture-specific boot code, a <code>printk</code> function for debugging, <code>malloc</code> for allocating memory and some low-level functions for talking to Xen.</dd>
  <dt>libocaml.a</dt>
  <dd>The OCaml runtime (the garbage collector, etc).</dd>
  <dt>libxencaml.a</dt>
  <dd>OCaml bindings for libminios and some boot code.</dd>
  <dt>libgcc.a</dt>
  <dd>Support functions for code that gcc generates (actually, not needed on x86).</dd>
</dl>

<p>To deploy the new unikernel, we create a Xen configuration file for it (here, I’m giving it 16 MB of RAM):</p>

<div><figure><figcaption><span>hw.xl </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
</pre></td>
<td><pre><code><span><span>name</span> <span>=</span> <span>'hw'</span>
</span><span><span>kernel</span> <span>=</span> <span>'hw.xen'</span>
</span><span><span>memory</span> <span>=</span> <span>16</span>
</span><span><span>on_crash</span> <span>=</span> <span>'preserve'</span>
</span><span><span>on_poweroff</span> <span>=</span> <span>'preserve'</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>Setting <code>on_crash</code> and <code>on_poweroff</code> to <code>preserve</code> lets us see any output or errors, which would otherwise be missed if the VM exits too quickly.</p>

<p>We can now boot our new VM:</p>

<pre><code>$ xl create -c hw.xl
Xen Minimal OS!
  start_info: 000000000009b000(VA)
    nr_pages: 0x800
  shared_inf: 0x6ee97000(MA)
     pt_base: 000000000009e000(VA)
nr_pt_frames: 0x5
    mfn_list: 0000000000097000(VA)
   mod_start: 0x0(VA)
     mod_len: 0
       flags: 0x0
    cmd_line: 
       stack: 0000000000055e00-0000000000075e00
Mirage: start_kernel
MM: Init
      _text: 0000000000000000(VA)
     _etext: 000000000003452d(VA)
   _erodata: 000000000003c000(VA)
     _edata: 000000000003e4d0(VA)
stack start: 0000000000055e00(VA)
       _end: 0000000000096d64(VA)
  start_pfn: a6
    max_pfn: 800
Mapping memory range 0x400000 - 0x800000
setting 0000000000000000-000000000003c000 readonly
skipped 0000000000001000
MM: Initialise page allocator for a8000(a8000)-800000(800000)
MM: done
Demand map pfns at 801000-2000801000.
Initialising timer interface
Initialising console ... done.
gnttab_table mapped at 0000000000801000.
xencaml: app_main_thread
getenv(OCAMLRUNPARAM) -&gt; null
getenv(CAMLRUNPARAM) -&gt; null
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Unsupported function lseek called in Mini-OS kernel
Hello, world!
main returned 0
</code></pre>

<p>( Note: I’m testing locally by running Xen under VirtualBox. Not all of Xen’s features can be used in this mode, but it works for testing unikernels. I’m also using my Git version of <code>mirage-xen</code>; the official one will display an error after printing the greeting because it expects you to provide a mainloop too. The warnings about <code>lseek</code> are just OCaml trying to find the current file offsets for <code>stdin</code>, <code>stdout</code> and <code>stderr</code>.)</p>

<p>As you can see, the boot process is quite short.
Execution begins at <a href="http://xenbits.xenproject.org/gitweb/?p=xen.git;a=blob;f=extras/mini-os/arch/x86/x86_64.S;h=df3469ef4319a75cc9d4c36b51f4097897c015f2;hb=HEAD#l19"><code>_start</code></a>.
Using <code>objdump -d hw.xen</code>, you can see that this just sets up the stack pointer register and calls the C function <code>arch_init</code>:</p>

<pre><code>0000000000000000 &lt;_start&gt;:
       0:   fc                      cld    
       1:   48 8b 25 0f 00 00 00    mov    0xf(%rip),%rsp        # 17 &lt;stack_start&gt;
       8:   48 81 e4 00 00 ff ff    and    $0xffffffffffff0000,%rsp
       f:   48 89 f7                mov    %rsi,%rdi
      12:   e8 e2 bb 00 00          callq  bbf9 &lt;arch_init&gt;
</code></pre>

<p><a href="http://xenbits.xenproject.org/gitweb/?p=xen.git;a=blob;f=extras/mini-os/arch/x86/setup.c;h=5e87dd1d99014a8139d8a63b375793661e57263b;hb=HEAD#l93">arch_init</a> (in libminios) initialises the traps and FPU and then prints <code>Xen Minimal OS!</code> and information about various addresses.
It then calls <code>start_kernel</code>.</p>

<p><a href="https://github.com/mirage/mirage-platform/blob/45814f85acca6174915acbb571146e1c4e978684/xen/runtime/xencaml/main.c#L57">start_kernel</a> (in libxencaml) sets up a few more features (events, interrupts, malloc, time-keeping and grant tables), then calls <code>caml_startup</code>.</p>

<p><a href="https://github.com/mirage/mirage-platform/blob/45814f85acca6174915acbb571146e1c4e978684/xen/runtime/ocaml/startup.c#L202">caml_startup</a> (in libocaml) initialises the garbage collector and calls <code>caml_program</code>, which is our <code>hw.native.o</code>.</p>

<p>We call <code>print_endline</code>, which libxencaml, as a convenience for debugging, forwards to libminios’s <code>console_print</code>.</p>

<h3 id="using-mirage-libraries">Using Mirage libraries</h3>

<p>The above was a bit of a hack, which ended up just using the C console driver in libminios (one of the few things it provides, as it’s needed for printk).
We can instead use the <code>mirage-console-xen</code> OCaml library, like this:</p>

<div><figure><figcaption><span>hw.ml </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
</pre></td>
<td><pre><code><span><span>open</span> <span>Lwt</span>
</span><span>
</span><span><span>let</span> <span>main</span> <span>=</span>
</span><span>  <span>Console</span><span>.</span><span>connect</span> <span>"0"</span> <span>&gt;&gt;=</span> <span>function</span>
</span><span>  <span>|</span> <span>`</span><span>Error</span> <span>_</span> <span>-&gt;</span> <span>failwith</span> <span>"Failed to connect to console"</span>
</span><span>  <span>|</span> <span>`</span><span>Ok</span> <span>default_console</span> <span>-&gt;</span>
</span><span>      <span>Console</span><span>.</span><span>log_s</span> <span>default_console</span> <span>"Hello, world!"</span>
</span><span>
</span><span><span>let</span> <span>()</span> <span>=</span>
</span><span>  <span>OS</span><span>.</span><span>Main</span><span>.</span><span>run</span> <span>main</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>Mirage uses the usual <code>Lwt</code> library for cooperative threading, which I wrote about at last year in <a href="http://roscidus.com/blog/blog/2013/11/28/asynchronous-python-vs-ocaml/">Asynchronous Python vs OCaml</a> - <code>&gt;&gt;=</code> means to wait for the result, allowing other code to run. Everything in Mirage is non-blocking, even looking up the console. <code>OS.Main.run</code> runs the main event loop.</p>

<p>Since we’re using libraries, let’s switch to ocamlbuild and give the dependencies in the <code>_tags</code> file, as usual for OCaml projects:</p>

<div><figure><figcaption><span>_tags </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
</pre></td>
<td><pre><code><span>true: warn(A), strict_sequence, package(mirage-console-xen)</span></code></pre></td>
</tr></table></div></figure></div>

<p>The only unusual thing we have to do here is tell ocamlbuild not to link in the <code>Unix</code> module when we build <code>hw.native.o</code>:</p>

<pre><code>$ ocamlbuild -lflags -linkpkg,-dontlink,unix -use-ocamlfind hw.native.o
</code></pre>

<p>In the same way, we can use other libraries to access raw block devices (<a href="https://github.com/mirage/mirage-block-xen">mirage-block-xen</a>), timers (<a href="https://github.com/mirage/mirage-clock">mirage-clock-xen</a>) and network interfaces (<a href="https://github.com/mirage/mirage-net-xen/">mirage-net-xen</a>).
Other (non-Xen-specific) OCaml libraries can then be used on top of these low-level drivers.
For example, <a href="https://github.com/mirage/ocaml-fat">fat-filesystem</a> can provide a filesystem on a block device, while <a href="https://github.com/mirage/mirage-tcpip">tcpip</a> provides an OCaml TCP/IP stack on a network interface.</p>

<h3 id="the-mirage-unix-libraries">The mirage-unix libraries</h3>

<p>You may have noticed that the Xen driver libraries we used above ended in <code>-xen</code>.
In fact, each of these is just an implementation of some generic interface provided by Mirage.
For example, <a href="https://github.com/mirage/mirage/blob/master/types/V1.mli">mirage/types</a> defines the abstract <code>CONSOLE</code> interface as:</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
<span>18</span>
<span>19</span>
<span>20</span>
<span>21</span>
<span>22</span>
<span>23</span>
<span>24</span>
<span>25</span>
<span>26</span>
<span>27</span>
<span>28</span>
<span>29</span>
<span>30</span>
<span>31</span>
</pre></td>
<td><pre><code><span><span>module</span> <span>type</span> <span>CONSOLE</span> <span>=</span> <span>sig</span>
</span><span>  <span>(** Text console input/output operations. *)</span>
</span><span>
</span><span>  <span>type</span> <span>error</span> <span>=</span> <span>[</span>
</span><span>    <span>|</span> <span>`</span><span>Invalid_console</span> <span>of</span> <span>string</span>
</span><span>  <span>]</span>
</span><span>  <span>(** The type representing possible errors when attaching a console. *)</span>
</span><span>
</span><span>  <span>include</span> <span>DEVICE</span> <span>with</span>
</span><span>    <span>type</span> <span>error</span> <span>:=</span> <span>error</span>
</span><span>
</span><span>  <span>val</span> <span>write</span> <span>:</span> <span>t</span> <span>-&gt;</span> <span>string</span> <span>-&gt;</span> <span>int</span> <span>-&gt;</span> <span>int</span> <span>-&gt;</span> <span>int</span>
</span><span>  <span>(** [write t buf off len] writes up to [len] chars of [String.sub buf</span>
</span><span><span>      off len] to the console [t] and returns the number of bytes</span>
</span><span><span>      written. Raises {!Invalid_argument} if [len &gt; buf - off]. *)</span>
</span><span>
</span><span>  <span>val</span> <span>write_all</span> <span>:</span> <span>t</span> <span>-&gt;</span> <span>string</span> <span>-&gt;</span> <span>int</span> <span>-&gt;</span> <span>int</span> <span>-&gt;</span> <span>unit</span> <span>io</span>
</span><span>  <span>(** [write_all t buf off len] is a thread that writes [String.sub buf</span>
</span><span><span>      off len] to the console [t] and returns when done. Raises</span>
</span><span><span>      {!Invalid_argument} if [len &gt; buf - off]. *)</span>
</span><span>
</span><span>  <span>val</span> <span>log</span> <span>:</span> <span>t</span> <span>-&gt;</span> <span>string</span> <span>-&gt;</span> <span>unit</span>
</span><span>  <span>(** [log str] writes as much characters of [str] that can be written</span>
</span><span><span>      in one write operation to the console [t], then writes</span>
</span><span><span>      "\r\n" to it. *)</span>
</span><span>
</span><span>  <span>val</span> <span>log_s</span> <span>:</span> <span>t</span> <span>-&gt;</span> <span>string</span> <span>-&gt;</span> <span>unit</span> <span>io</span>
</span><span>  <span>(** [log_s str] is a thread that writes [str ^ "\r\n"] in the</span>
</span><span><span>      console [t]. *)</span>
</span><span>
</span><span><span>end</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>By linking against the <code>-unix</code> versions of libraries rather than the <code>-xen</code> ones, we can compile our code as an ordinary Unix program and run it directly.
This makes testing and debugging very easy.</p>

<p>To make sure our code is generic enough to do this, we can wrap it in a functor that takes any console module as an input:</p>

<div><figure><figcaption><span>unikernel.ml </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
</pre></td>
<td><pre><code><span><span>module</span> <span>Main</span> <span>(</span><span>C</span> <span>:</span> <span>V1_LWT</span><span>.</span><span>CONSOLE</span><span>)</span> <span>=</span> <span>struct</span>
</span><span>  <span>let</span> <span>start</span> <span>c</span> <span>=</span>
</span><span>    <span>C</span><span>.</span><span>log_s</span> <span>c</span> <span>"Hello, world!"</span>
</span><span><span>end</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>The code that provides a Xen or Unix console and calls this goes in <code>main.ml</code>:</p>

<div><figure><figcaption><span>main.ml </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
</pre></td>
<td><pre><code><span><span>open</span> <span>Lwt</span>
</span><span>
</span><span><span>let</span> <span>console</span> <span>=</span>
</span><span>  <span>Console</span><span>.</span><span>connect</span> <span>"0"</span> <span>&gt;&gt;=</span> <span>function</span>
</span><span>  <span>|</span> <span>`</span><span>Error</span> <span>_</span> <span>-&gt;</span> <span>failwith</span> <span>"Failed to connect to console"</span>
</span><span>  <span>|</span> <span>`</span><span>Ok</span> <span>c</span> <span>-&gt;</span> <span>return</span> <span>c</span>
</span><span>
</span><span><span>module</span> <span>U</span> <span>=</span> <span>Unikernel</span><span>.</span><span>Main</span><span>(</span><span>Console</span><span>)</span>
</span><span>
</span><span><span>let</span> <span>()</span> <span>=</span>
</span><span>  <span>OS</span><span>.</span><span>Main</span><span>.</span><span>run</span> <span>(</span><span>console</span> <span>&gt;&gt;=</span> <span>U</span><span>.</span><span>start</span><span>)</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>With the platform-specific code isolated in <code>main.ml</code>, we can now use the <code>mirage</code> command-line tool to generate it automatically for the target platform.
<code>mirage</code> takes a <code>config.ml</code> configuration file and generates <code>Makefile</code> and <code>main.ml</code> based on the current platform and the arguments passed.</p>

<div><figure><figcaption><span>config.ml </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
</pre></td>
<td><pre><code><span><span>open</span> <span>Mirage</span>
</span><span>
</span><span><span>let</span> <span>main</span> <span>=</span> <span>foreign</span> <span>"Unikernel.Main"</span> <span>(</span><span>console</span> <span>@-&gt;</span> <span>job</span><span>)</span>
</span><span>
</span><span><span>let</span> <span>()</span> <span>=</span>
</span><span>  <span>register</span> <span>"hw"</span> <span>[</span>
</span><span>    <span>main</span> <span>$</span> <span>default_console</span>
</span><span>  <span>]</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<pre><code>$ mirage configure --unix
$ make
$ ./mir-hw 
Hello, world!
</code></pre>

<p>I won’t describe this in detail because at this point we’ve reached the start of the <a href="http://openmirage.org/wiki/hello-world">official tutorial</a>, and you can read that instead.</p>

<h2 id="test-case">Test case</h2>

<p>Because <a href="http://0install.net">0install</a> is decentralised, it doesn’t need a single centrally-managed repository (or several incompatible repositories, each trying to package every program, as is common with Linux distributions).
In 0install, it’s possible for every developer to run their own <a href="http://0install.net/0repo.html">repository</a>, containing just their software, with cross-repository dependencies handled automatically.
But just because it’s possible doesn’t mean we have to go to that extreme: having medium sized repositories each managed by a team of people can be very convenient, especially where package maintainers come and go.</p>

<p>The general pattern for a group repository is to have a public server that accepts new package uploads from developers, and a private (firewalled) server with the repository’s GPG key, which downloads from it:</p>

<p><img src="http://roscidus.com/blog/images/0repo-multi.png"></p>

<p>Debian uses an anonymous FTP server for its incoming queue, polling it with a cron job.
This turns out to be surprisingly complicated.
You need to handle incomplete uploads (not processing them until they’re done, or deleting them eventually if they never complete), allow contributors to overwrite or delete their own partial uploads (Debian allows you to upload a GPG-signed command file, which provides some control), etc, as well as keep the service fully patched.
Also, the cron system can be annoying: if the package contains a mistake then it will be several minutes before it discovers this and emails the packager.</p>

<p>Perhaps there are some decent systems out there to handle all this, but it seemed like a good opportunity to try making a unikernel.</p>

<p>A particularly nice feature of this test-case is that it doesn’t matter too much if it fails:
the repository itself will check the developer’s signature on the files, so an attacker can’t compromise the repository by breaking into the queue; everything in the queue is intended to become public, so we need not worry much about confidentiality; lost uploads can be easily resubmitted; and if it goes down for a bit, it just means that new software can’t be added to the repository.
So, there’s nothing critical about this service, which is reassuring.</p>

<h3 id="storage">Storage</h3>

<p>The <a href="https://github.com/mirage/merge-queues">merge-queues</a> library builds a queue abstraction on top of
<a href="https://github.com/mirage/irmin">Irmin</a>, a Git-inspired storage system for Mirage.
But my needs are simple, and I wanted to test the more primitive libraries first, so I decided to build my queue directly on a plain filesystem.
This was the first interface I came up with:</p>

<div><figure><figcaption><span>upload_queue.mli </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
<span>18</span>
<span>19</span>
<span>20</span>
<span>21</span>
<span>22</span>
<span>23</span>
<span>24</span>
<span>25</span>
<span>26</span>
<span>27</span>
<span>28</span>
<span>29</span>
<span>30</span>
<span>31</span>
<span>32</span>
<span>33</span>
<span>34</span>
<span>35</span>
<span>36</span>
<span>37</span>
<span>38</span>
<span>39</span>
<span>40</span>
<span>41</span>
<span>42</span>
<span>43</span>
<span>44</span>
<span>45</span>
<span>46</span>
<span>47</span>
<span>48</span>
</pre></td>
<td><pre><code><span><span>module</span> <span>type</span> <span>FS</span> <span>=</span> <span>V1_LWT</span><span>.</span><span>FS</span> <span>with</span>
</span><span>  <span>type</span> <span>page_aligned_buffer</span> <span>=</span> <span>Cstruct</span><span>.</span><span>t</span> <span>and</span>
</span><span>  <span>type</span> <span>block_device_error</span> <span>=</span> <span>Fat</span><span>.</span><span>Fs</span><span>.</span><span>block_error</span>
</span><span>
</span><span><span>(** An upload.</span>
</span><span><span> * To avoid loading complete uploads into RAM, we stream them</span>
</span><span><span> * between the network and the disk. *)</span>
</span><span><span>type</span> <span>item</span> <span>=</span> <span>{</span>
</span><span>  <span>size</span> <span>:</span> <span>int64</span><span>;</span>
</span><span>  <span>data</span> <span>:</span> <span>string</span> <span>Lwt_stream</span><span>.</span><span>t</span><span>;</span>
</span><span><span>}</span>
</span><span>
</span><span><span>type</span> <span>add_error</span> <span>=</span> <span>[`</span><span>Wrong_size</span> <span>of</span> <span>int64</span> <span>|</span> <span>`</span><span>Unknown</span> <span>of</span> <span>exn</span><span>]</span>
</span><span>
</span><span><span>module</span> <span>Make</span> <span>:</span> <span>functor</span> <span>(</span><span>F</span> <span>:</span> <span>FS</span><span>)</span> <span>-&gt;</span> <span>sig</span>
</span><span>  <span>(** An upload queue. *)</span>
</span><span>  <span>type</span> <span>t</span>
</span><span>
</span><span>  <span>(** Create a new queue, backed by a filesystem. *)</span>
</span><span>  <span>val</span> <span>create</span> <span>:</span> <span>F</span><span>.</span><span>t</span> <span>-&gt;</span> <span>t</span> <span>Lwt</span><span>.</span><span>t</span>
</span><span>
</span><span>  <span>module</span> <span>Upload</span> <span>:</span> <span>sig</span>
</span><span>    <span>(** Add an upload to the queue.</span>
</span><span><span>     * The upload is added only once the end of the stream is</span>
</span><span><span>     * reached, and only if the total size matches the size</span>
</span><span><span>     * in the record.</span>
</span><span><span>     * To cancel an add, just terminate the stream. *)</span>
</span><span>    <span>val</span> <span>add</span> <span>:</span>
</span><span>      <span>t</span> <span>-&gt;</span> <span>item</span> <span>-&gt;</span> <span>[</span> <span>`</span><span>Ok</span> <span>of</span> <span>unit</span> <span>|</span> <span>`</span><span>Error</span> <span>of</span> <span>add_error</span> <span>]</span> <span>Lwt</span><span>.</span><span>t</span>
</span><span>  <span>end</span>
</span><span>
</span><span>  <span>module</span> <span>Download</span> <span>:</span> <span>sig</span>
</span><span>    <span>(** Interface for the repository software to fetch items</span>
</span><span><span>     * from the queue. Only one client may use this interface</span>
</span><span><span>     * at a time, or things will go wrong. *)</span>
</span><span>
</span><span>    <span>(** Return a fresh stream for the item at the head of the</span>
</span><span><span>     * queue, without removing it. After downloading it</span>
</span><span><span>     * successfully, the client should call [delete]. If the</span>
</span><span><span>     * queue is empty, this blocks until an item is available. *)</span>
</span><span>    <span>val</span> <span>peek</span> <span>:</span> <span>t</span> <span>-&gt;</span> <span>item</span> <span>Lwt</span><span>.</span><span>t</span>
</span><span>
</span><span>    <span>(** Delete the item previously retrieved by [peek].</span>
</span><span><span>     * If the previous item has already been deleted, this does</span>
</span><span><span>     * nothing, even if there are more items in the queue. *)</span>
</span><span>    <span>val</span> <span>delete</span> <span>:</span> <span>t</span> <span>-&gt;</span> <span>unit</span> <span>Lwt</span><span>.</span><span>t</span>
</span><span>  <span>end</span>
</span><span><span>end</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>Our <code>unikernel.ml</code> will use this to make a queue, backed by a filesystem.
Uploaders’ HTTP POSTs will be routed to <code>Upload.add</code>, while the repository’s GET and DELETE invocations go to the <code>Download</code> submodule.
<code>delete</code> is a separate operation because we want the repository to confirm that it got the item successfully before we delete it, in case of network errors.</p>

<p>Ideally, we might require that the <code>DELETE</code> comes over the same HTTP connection as the <code>GET</code> just in case we accidentally run two instances of the repository software, but that’s unlikely and it’s convenient to test using separate <code>curl</code> invocations.</p>

<p>We’re using another functor here, <code>Upload_queue.Make</code>, so that our queue will work over any filesystem.
In theory, we can configure our unikernel with a FAT filesystem on a block device when running under Xen,
while using a regular directory when running under Linux (e.g. for testing).</p>

<p>But it doesn’t work.
You can see at the top that I had to restrict Mirage’s abstract <code>FS</code> type in two ways:</p>

<ul>
  <li>
    <p>The <code>read</code> and <code>write</code> functions in <code>FS</code> pass the data using the abstract <code>page_aligned_buffer</code> type.
Since we need to do something with the data, this isn’t good enough.
I therefore declare that this must be a <code>Cstruct.t</code> (basically, an array of bytes).
This is actually OK; <a href="https://github.com/mirage/mirage-fs-unix">mirage-fs-unix</a> also uses this type.</p>
  </li>
  <li>
    <p>One of the possible error codes from <code>FS</code> is the abstract type <code>FS.block_device_error</code>, and I can’t
see any way to turn one of these into a string using the <code>FS</code> interface.
I therefore require a filesystem implementation that defines it to be <code>Fat.Fs.block_error</code>.
Obviously, this means we now only support the FAT filesystem.</p>
  </li>
</ul>

<p>This doesn’t prevent us from running as a normal process, because we can ask for a Unix “block” device (actually, just a plain <code>disk.img</code> file) and pass that to the <code>Fat</code> module, but it would be nice to have the option of using a real directory.</p>

<p>I asked about this on the mailing list - <a href="http://lists.xenproject.org/archives/html/mirageos-devel/2014-07/msg00048.html">Mirage questions from writing a REST service</a> - and it looks like the <code>FS</code> type will change soon.</p>

<h3 id="implementation">Implementation</h3>

<p>For the curious, this initial implementation is in <a href="https://github.com/0install/0repo-queue/blob/063db53c6faf76c7a9edd18416c25459603b777d/upload_queue.ml">upload_queue.ml</a>.</p>

<p>Internally, the module creates an in-memory queue to keep track of successful uploads.
Uploads are streamed to the disk and
when an upload completes with the declared size, the filename is added to the queue.
If the upload ends with the wrong size (probably because the connection was lost), the file is deleted.</p>

<p>But what if our VM gets rebooted?
We need to scan the file system at start up and work out which uploads are complete and which should be deleted.
My first thought was to name the files <code>NUMBER.part</code> during the upload and rename on success.
However, the <code>FS</code> interface currently lacks a <code>rename</code> method.
Instead, I write an <code>N</code> byte to the start of each file and set it to <code>Y</code> on success.
That works, but renaming would be nicer!</p>

<p>For downloading, the <code>peek</code> function returns the item at the head of the queue.
If the queue is empty, it waits until something arrives.
The repository just makes a GET request - if something is available then it returns immediately,
otherwise the connection stays open until some data is ready, allowing the repository to respond immediately to new uploads.</p>

<h3 id="unit-testing-the-storage-system">Unit-testing the storage system</h3>

<p>Because our unikernel can run as a process, testing is easy even if you don’t have a local Xen deployment.
A set of unit-tests test the upload queue module just as for any other program, and the service can be run as a normal process, listening on a normal TCP socket.
A slight annoyance here is that the generated Makefile doesn’t include any rules to build the tests so you have to add them manually, and
if you regenerate the Makefile then it loses the new rule.</p>

<p>As you might expect from such a new system, testing uncovered several problems. The first (minor) problem is that when the disk becomes full, the unhelpful error reported by the filesystem is <code>Failure("Unknown error: Failure(\"fault\")")</code>.</p>

<p>( I asked about this on the mailing list - <a href="http://lists.xenproject.org/archives/html/mirageos-devel/2014-07/msg00069.html">Error handling in Mirage</a> - and there seems to be agreement that error handling should change. )</p>

<p>A more serious problem was that deleting files corrupted the FAT directory index.
I downloaded the FAT library and added a unit-test for delete, which made it easy to track the problem down (despite my lack of knowledge of FAT).
Here’s the code for marking a directory entry as deleted in the FAT library:</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
</pre></td>
<td><pre><code><span>    <span>let</span> <span>b</span> <span>=</span> <span>Cstruct</span><span>.</span><span>sub</span> <span>block</span> <span>offset</span> <span>sizeof</span> <span>in</span>
</span><span>    <span>let</span> <span>delta</span> <span>=</span> <span>Cstruct</span><span>.</span><span>create</span> <span>sizeof</span> <span>in</span>
</span><span>    <span>begin</span> <span>match</span> <span>unmarshal</span> <span>b</span> <span>with</span>
</span><span>      <span>|</span> <span>Lfn</span> <span>lfn</span> <span>-&gt;</span>
</span><span>	<span>let</span> <span>lfn'</span> <span>=</span> <span>{</span> <span>lfn</span> <span>with</span> <span>lfn_deleted</span> <span>=</span> <span>true</span> <span>}</span> <span>in</span>
</span><span>	<span>marshal</span> <span>delta</span> <span>(</span><span>Lfn</span> <span>lfn'</span><span>)</span>
</span><span>      <span>|</span> <span>Dos</span> <span>dos</span> <span>-&gt;</span>
</span><span>	<span>let</span> <span>dos'</span> <span>=</span> <span>{</span> <span>dos</span> <span>with</span> <span>deleted</span> <span>=</span> <span>true</span> <span>}</span> <span>in</span>
</span><span>	<span>marshal</span> <span>b</span> <span>(</span><span>Dos</span> <span>dos'</span><span>)</span>
</span><span>      <span>|</span> <span>End</span> <span>-&gt;</span> <span>assert</span> <span>false</span>
</span><span>    <span>end</span><span>;</span>
</span><span>    <span>Update</span><span>.</span><span>from_cstruct</span> <span>(</span><span>Int64</span><span>.</span><span>of_int</span> <span>offset</span><span>)</span> <span>delta</span> <span>::</span> <span>acc</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>It’s supposed to take an entry, unmarshal it into an OCaml structure, set the <code>deleted</code> flag, and marshal the result into a new <code>delta</code> structure.
These deltas are returned and applied to the device.
The bug is a simple typo: <code>Lfn</code> (long filename) entries update correctly, but for old <code>Dos</code> ones it writes the new block to the input, not to <code>delta</code>.
The fix was simple enough (I also refactored it slightly to encourage the correct behaviour in future):</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
</pre></td>
<td><pre><code><span>    <span>let</span> <span>b</span> <span>=</span> <span>Cstruct</span><span>.</span><span>sub</span> <span>block</span> <span>offset</span> <span>sizeof</span> <span>in</span>
</span><span>    <span>let</span> <span>delta</span> <span>=</span> <span>Cstruct</span><span>.</span><span>create</span> <span>sizeof</span> <span>in</span>
</span><span>    <span>marshal</span> <span>delta</span> <span>begin</span> <span>match</span> <span>unmarshal</span> <span>b</span> <span>with</span>
</span><span>      <span>|</span> <span>Lfn</span> <span>lfn</span> <span>-&gt;</span> <span>Lfn</span> <span>{</span> <span>lfn</span> <span>with</span> <span>lfn_deleted</span> <span>=</span> <span>true</span> <span>}</span>
</span><span>      <span>|</span> <span>Dos</span> <span>dos</span> <span>-&gt;</span> <span>Dos</span> <span>{</span> <span>dos</span> <span>with</span> <span>deleted</span> <span>=</span> <span>true</span> <span>}</span>
</span><span>      <span>|</span> <span>End</span> <span>-&gt;</span> <span>assert</span> <span>false</span>
</span><span>    <span>end</span><span>;</span>
</span><span>    <span>Update</span><span>.</span><span>from_cstruct</span> <span>(</span><span>Int64</span><span>.</span><span>of_int</span> <span>offset</span><span>)</span> <span>delta</span> <span>::</span> <span>acc</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>This demonstrates both the good and the bad of Mirage: the bug was easy to find and fix, using regular debugging tools.
I’m sure fixing a filesystem corruption bug in the Linux kernel would have been vastly more difficult.
On the other hard, Linux is rather well tested, whereas I appear to be the first person ever to try deleting a file in Mirage!</p>

<h3 id="the-http-server">The HTTP server</h3>

<p>This turned out to be quite simple. Here’s the unikernel’s <a href="https://github.com/0install/0repo-queue/blob/92c99d34104b6bea707044924f9753a9cc7f0414/unikernel.ml"><code>start</code></a> function:</p>

<div><figure><figcaption><span>unikernel.ml </span></figcaption>
<div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
<span>18</span>
<span>19</span>
<span>20</span>
<span>21</span>
<span>22</span>
<span>23</span>
<span>24</span>
<span>25</span>
<span>26</span>
<span>27</span>
<span>28</span>
<span>29</span>
</pre></td>
<td><pre><code><span><span>module</span> <span>Main</span> <span>(</span><span>C</span> <span>:</span> <span>V1_LWT</span><span>.</span><span>CONSOLE</span><span>)</span>
</span><span>            <span>(</span><span>F</span> <span>:</span> <span>Upload_queue</span><span>.</span><span>FS</span><span>)</span>
</span><span>            <span>(</span><span>H</span> <span>:</span> <span>Cohttp_lwt</span><span>.</span><span>Server</span><span>)</span> <span>=</span> <span>struct</span>
</span><span>  <span>module</span> <span>Q</span> <span>=</span> <span>Upload_queue</span><span>.</span><span>Make</span><span>(</span><span>F</span><span>)</span>
</span><span>  <span>[...]</span>
</span><span>  <span>let</span> <span>start</span> <span>c</span> <span>fs</span> <span>http</span> <span>=</span>
</span><span>    <span>Log</span><span>.</span><span>write</span> <span>:=</span> <span>C</span><span>.</span><span>log_s</span> <span>c</span><span>;</span>
</span><span>    <span>Log</span><span>.</span><span>info</span> <span>"starting queue service"</span> <span>&gt;&gt;=</span> <span>fun</span> <span>()</span> <span>-&gt;</span>
</span><span>
</span><span>    <span>Q</span><span>.</span><span>create</span> <span>fs</span> <span>&gt;&gt;=</span> <span>fun</span> <span>q</span> <span>-&gt;</span>
</span><span>
</span><span>    <span>let</span> <span>callback</span> <span>_</span><span>conn_id</span> <span>request</span> <span>body</span> <span>=</span>
</span><span>      <span>match</span> <span>Uri</span><span>.</span><span>path</span> <span>request</span><span>.</span><span>H</span><span>.</span><span>Request</span><span>.</span><span>uri</span> <span>with</span>
</span><span>      <span>|</span> <span>"/uploader"</span> <span>-&gt;</span> <span>handle_uploader</span> <span>q</span> <span>request</span> <span>body</span>
</span><span>      <span>|</span> <span>"/downloader"</span> <span>-&gt;</span> <span>handle_downloader</span> <span>q</span> <span>request</span>
</span><span>      <span>|</span> <span>path</span> <span>-&gt;</span>
</span><span>          <span>H</span><span>.</span><span>respond_error</span>
</span><span>	    <span>~</span><span>status</span><span>:`</span><span>Bad_request</span>
</span><span>      	    <span>~</span><span>body</span><span>:(</span><span>Printf</span><span>.</span><span>sprintf</span> <span>"Bad path '%s'</span><span>\n</span><span>"</span> <span>path</span><span>)</span>
</span><span>	    <span>()</span> <span>in</span>
</span><span>
</span><span>    <span>let</span> <span>conn_closed</span> <span>_</span><span>conn_id</span> <span>()</span> <span>=</span>
</span><span>      <span>Log</span><span>.</span><span>info</span> <span>"connection closed"</span> <span>|&gt;</span> <span>ignore</span> <span>in</span>
</span><span>
</span><span>    <span>http</span> <span>{</span> <span>H</span><span>.</span>
</span><span>      <span>callback</span><span>;</span>
</span><span>      <span>conn_closed</span>
</span><span>    <span>}</span>
</span><span><span>end</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>Here, our functor is extended to take a filesystem (using the restricted type required by our <code>Upload_queue</code>, as noted above) and an HTTP server module as arguments.</p>

<p>The HTTP server calls our <code>callback</code> each time it receives a request, and this dispatches <code>/uploader</code> requests to <code>handle_uploader</code> and <code>/downloader</code> ones to <code>handle_downloader</code>. These are also very simple, e.g.</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
</pre></td>
<td><pre><code><span>  <span>let</span> <span>get</span> <span>q</span> <span>=</span>
</span><span>    <span>Q</span><span>.</span><span>Download</span><span>.</span><span>peek</span> <span>q</span> <span>&gt;&gt;=</span> <span>fun</span> <span>{</span><span>Upload_queue</span><span>.</span><span>size</span><span>;</span> <span>data</span><span>}</span> <span>-&gt;</span>
</span><span>    <span>let</span> <span>body</span> <span>=</span> <span>Cohttp_lwt_body</span><span>.</span><span>of_stream</span> <span>data</span> <span>in</span>
</span><span>    <span>let</span> <span>headers</span> <span>=</span> <span>Cohttp</span><span>.</span><span>Header</span><span>.</span><span>init_with</span>
</span><span>      <span>"Content-Length"</span> <span>(</span><span>Int64</span><span>.</span><span>to_string</span> <span>size</span><span>)</span> <span>in</span>
</span><span>    <span>(* Adding a content-length loses the transfer-encoding</span>
</span><span><span>     * for some reason, so add it back: *)</span>
</span><span>    <span>let</span> <span>headers</span> <span>=</span> <span>Cohttp</span><span>.</span><span>Header</span><span>.</span><span>add</span> <span>headers</span>
</span><span>      <span>"transfer-encoding"</span> <span>"chunked"</span> <span>in</span>
</span><span>    <span>H</span><span>.</span><span>respond</span> <span>~</span><span>headers</span> <span>~</span><span>status</span><span>:`</span><span>OK</span> <span>~</span><span>body</span> <span>()</span>
</span><span>
</span><span>  <span>let</span> <span>handle_downloader</span> <span>q</span> <span>request</span> <span>=</span>
</span><span>    <span>match</span> <span>H</span><span>.</span><span>Request</span><span>.</span><span>meth</span> <span>request</span> <span>with</span>
</span><span>    <span>|</span> <span>`</span><span>GET</span> <span>-&gt;</span> <span>get</span> <span>q</span>
</span><span>    <span>|</span> <span>`</span><span>DELETE</span> <span>-&gt;</span> <span>delete</span> <span>q</span>
</span><span>    <span>|</span> <span>`</span><span>HEAD</span> <span>|</span> <span>`</span><span>PUT</span> <span>|</span> <span>`</span><span>POST</span>
</span><span>    <span>|</span> <span>`</span><span>OPTIONS</span> <span>|</span> <span>`</span><span>PATCH</span> <span>-&gt;</span> <span>unsupported_method</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>The other methods (<code>put</code> and <code>delete</code>) are similar.</p>

<h3 id="buffered-reads">Buffered reads</h3>

<p>Running as a <code>--unix</code> process, I initially got a download speed of
17.2 KB/s, which was rather disappointing.
Especially as Apache on the same machine gets 615 MB/s!</p>

<p>Increasing the size of the chunks I was reading from the Fat
filesystem (a disk.img file) from 512 bytes to 1MB, I was able to
increase this to 2.83 MB/s, and removing the <code>O_DIRECT</code> flag from
<code>mirage-block-unix</code>, download speed increased to 15 MB/s (so this is
with Linux caching the data in RAM).</p>

<p>To check the filesystem was the problem, I removed the <code>F.read</code> call
(so it would return uninitialised data instead of the actual file contents).
It then managed a very respectable 514 MB/s.
Nothing wrong with the HTTP code then.</p>

<h3 id="streaming-uploads">Streaming uploads</h3>

<p>It all worked nicely running as a Unix process, so the next step was to deploy on Xen.
I was hoping that most of the bugs would already have been found during the Unix testing,
but in fact there were more lurking.</p>

<p>It worked for very small files, but when uploading larger files it quickly ran
out of memory on my 64-bit x86 test system. I also tried it on my 32-bit CubieTruck
ARM board, but that failed even sooner, with <code>Invalid_argument("String.create")</code> (on 32-bit
platforms, OCaml strings are limited to 16 MB).</p>

<p>In both cases, the problem was that the <a href="https://github.com/mirage/ocaml-cohttp">cohttp</a> library tried to read the entire upload in one go.
I found the <a href="https://github.com/mirage/ocaml-cohttp/blob/86394acdb580257ee78cb6976966662575b01bb0/cohttp/transfer_io.ml#L56">read</a> function in <code>Transfer_io</code>:</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
</pre></td>
<td><pre><code><span><span>let</span> <span>read</span> <span>~</span><span>len</span> <span>ic</span> <span>=</span>
</span><span>  <span>(* TODO functorise string to a bigbuffer *)</span>
</span><span>  <span>match</span> <span>len</span> <span>with</span>
</span><span>  <span>|</span><span>0</span> <span>-&gt;</span> <span>return</span> <span>Done</span>
</span><span>  <span>|</span><span>len</span> <span>-&gt;</span>
</span><span>    <span>read_exactly</span> <span>ic</span> <span>len</span> <span>&gt;&gt;=</span> <span>function</span>
</span><span>    <span>|</span><span>None</span> <span>-&gt;</span> <span>return</span> <span>Done</span>
</span><span>    <span>|</span><span>Some</span> <span>buf</span> <span>-&gt;</span> <span>return</span> <span>(</span><span>Final_chunk</span> <span>buf</span><span>)</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>I changed it to use <code>read</code> rather than <code>read_exactly</code> (<code>read</code> returns whatever data is available, waiting only if there isn’t any at all):</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
</pre></td>
<td><pre><code><span><span>let</span> <span>read</span> <span>~</span><span>remaining</span> <span>ic</span> <span>=</span>
</span><span>  <span>(* TODO functorise string to a bigbuffer *)</span>
</span><span>  <span>match</span> <span>!</span><span>remaining</span> <span>with</span>
</span><span>  <span>|</span><span>0</span> <span>-&gt;</span> <span>return</span> <span>Done</span>
</span><span>  <span>|</span><span>len</span> <span>-&gt;</span>
</span><span>    <span>read</span> <span>ic</span> <span>len</span> <span>&gt;&gt;=</span> <span>fun</span> <span>buf</span> <span>-&gt;</span>
</span><span>    <span>remaining</span> <span>:=</span> <span>!</span><span>remaining</span> <span>-</span> <span>String</span><span>.</span><span>length</span> <span>buf</span><span>;</span>
</span><span>    <span>if</span> <span>!</span><span>remaining</span> <span>=</span> <span>0</span> <span>then</span> <span>return</span> <span>(</span><span>Final_chunk</span> <span>buf</span><span>)</span>
</span><span>    <span>else</span> <span>return</span> <span>(</span><span>Chunk</span> <span>buf</span><span>)</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>I also had to change the signature to take a mutable reference (<code>remaining</code>) for the remaining data, otherwise it has no way to know when it’s done (<a href="https://github.com/talex5/ocaml-cohttp/commit/030969c32d310d690b8df5929d9a5f32caee6eb5">patch</a>).</p>

<h3 id="buffered-writes">Buffered writes</h3>

<p>With the uploads now split into chunks, upload speed with <code>--unix</code> was 178 KB/s.
Batching up the chunks (which were generally 4 KB each) into a 64 KB buffer increased the speed to 2083 KB/s.
With a 1 MB buffer, I got 6386 KB/s.</p>

<p>Here’s the code I used:</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
<span>18</span>
<span>19</span>
<span>20</span>
<span>21</span>
<span>22</span>
<span>23</span>
<span>24</span>
<span>25</span>
<span>26</span>
<span>27</span>
<span>28</span>
<span>29</span>
<span>30</span>
<span>31</span>
<span>32</span>
<span>33</span>
</pre></td>
<td><pre><code><span><span>let</span> <span>page_buffer</span> <span>=</span> <span>Io_page</span><span>.</span><span>get</span> <span>256</span> <span>|&gt;</span> <span>Io_page</span><span>.</span><span>to_cstruct</span> <span>in</span>
</span><span>
</span><span><span>(* Set the first byte to N to indicate that we're not done yet.</span>
</span><span><span> * If we reboot while this flag is set, the partial upload will</span>
</span><span><span> * be deleted. *)</span>
</span><span><span>let</span> <span>page_buffer_used</span> <span>=</span> <span>ref</span> <span>1</span> <span>in</span>
</span><span><span>Cstruct</span><span>.</span><span>set_char</span> <span>page_buffer</span> <span>0</span> <span>'N'</span><span>;</span>
</span><span><span>let</span> <span>file_offset</span> <span>=</span> <span>ref</span> <span>1</span> <span>in</span>
</span><span>
</span><span><span>let</span> <span>flush_page_buffer</span> <span>()</span> <span>=</span>
</span><span>  <span>Log</span><span>.</span><span>info</span> <span>"Flushing %d bytes to disk"</span> <span>!</span><span>page_buffer_used</span> <span>&gt;&gt;=</span> <span>fun</span> <span>()</span> <span>-&gt;</span>
</span><span>  <span>let</span> <span>buffered_data</span> <span>=</span> <span>Cstruct</span><span>.</span><span>sub</span> <span>page_buffer</span> <span>0</span> <span>!</span><span>page_buffer_used</span> <span>in</span>
</span><span>  <span>F</span><span>.</span><span>write</span> <span>q</span><span>.</span><span>fs</span> <span>name</span> <span>!</span><span>file_offset</span> <span>buffered_data</span> <span>&gt;&gt;|=</span> <span>fun</span> <span>()</span> <span>-&gt;</span>
</span><span>  <span>file_offset</span> <span>:=</span> <span>!</span><span>file_offset</span> <span>+</span> <span>!</span><span>page_buffer_used</span><span>;</span>
</span><span>  <span>page_buffer_used</span> <span>:=</span> <span>0</span><span>;</span>
</span><span>  <span>return</span> <span>()</span> <span>in</span>
</span><span>
</span><span><span>let</span> <span>rec</span> <span>add_data</span> <span>src</span> <span>i</span> <span>=</span>
</span><span>  <span>let</span> <span>src_remaining</span> <span>=</span> <span>String</span><span>.</span><span>length</span> <span>src</span> <span>-</span> <span>i</span> <span>in</span>
</span><span>  <span>if</span> <span>src_remaining</span> <span>=</span> <span>0</span> <span>then</span> <span>return</span> <span>()</span>
</span><span>  <span>else</span> <span>(</span>
</span><span>    <span>let</span> <span>page_buffer_free</span> <span>=</span> <span>Cstruct</span><span>.</span><span>len</span> <span>page_buffer</span> <span>-</span> <span>!</span><span>page_buffer_used</span> <span>in</span>
</span><span>    <span>let</span> <span>chunk_size</span> <span>=</span> <span>min</span> <span>page_buffer_free</span> <span>src_remaining</span> <span>in</span>
</span><span>    <span>Cstruct</span><span>.</span><span>blit_from_string</span> <span>src</span> <span>i</span> <span>page_buffer</span> <span>!</span><span>page_buffer_used</span> <span>chunk_size</span><span>;</span>
</span><span>    <span>page_buffer_used</span> <span>:=</span> <span>!</span><span>page_buffer_used</span> <span>+</span> <span>chunk_size</span><span>;</span>
</span><span>    <span>lwt</span> <span>()</span> <span>=</span>
</span><span>      <span>if</span> <span>page_buffer_free</span> <span>=</span> <span>chunk_size</span> <span>then</span> <span>flush_page_buffer</span> <span>()</span>
</span><span>      <span>else</span> <span>return</span> <span>()</span> <span>in</span>
</span><span>    <span>add_data</span> <span>src</span> <span>(</span><span>i</span> <span>+</span> <span>chunk_size</span><span>)</span>
</span><span>  <span>)</span> <span>in</span>
</span><span>
</span><span><span>data</span> <span>|&gt;</span> <span>Lwt_stream</span><span>.</span><span>iter_s</span> <span>(</span><span>fun</span> <span>data</span> <span>-&gt;</span> <span>add_data</span> <span>data</span> <span>0</span><span>)</span> <span>&gt;&gt;=</span>
</span><span><span>flush_page_buffer</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>Asking on the mailing list confirmed that
<a href="http://lists.xenproject.org/archives/html/mirageos-devel/2014-07/msg00054.html">Fat is not well optimised</a>.
This isn’t actually a problem for my service, since it’s still faster than
my Internet connection, but there’s clearly more work needed here.</p>

<h3 id="upload-speed-on-xen">Upload speed on Xen</h3>

<p>Testing on my little CubieTruck board, I then got:</p>

<table>
  <tbody>
    <tr>
      <td>Upload speed</td>
      <td>74 KB/s</td>
    </tr>
    <tr>
      <td>Download speed</td>
      <td>1.6 KB/s</td>
    </tr>
  </tbody>
</table>

<p>Hmm. To get a feel for what the board is capable of, I ran <code>nc -l -p 8080 &lt; /dev/zero</code> on the board
and <code>nc cubietruck 8080 | pv &gt; /dev/null</code> on my laptop, getting 29 MB/s.</p>

<p>Still, my unikernel is running as a guest, meaning it has the overhead of using the virtual network
interface (it has to pass the data to dom0, which then sends it over the real interface). So I installed
a Linux guest and tried from there. 47.2 MB/s. Interesting. I have no idea why it’s faster than dom0!</p>

<p>I loaded up <a href="http://www.wireshark.org/">Wireshark</a> to see what was happening with the unikernel transfers.
The upload transfer mostly went fast, but stalled in the middle for 15 seconds and then for 12 seconds at the end.
Wireshark showed that the unikernel was ack’ing the packets but reducing the TCP window size, indicating that the packets weren’t being processed by the application code.
The delays corresponded to the times when we were flushing the data to the SD card, which makes sense.
So, this looks like another filesystem problem (we should be able to write to the SD card much faster than this).</p>

<h3 id="tcp-retransmissions">TCP retransmissions</h3>

<p>For the download, Wireshark showed that many of the packets had incorrect TCP checksums and were having to be retransmitted.
I was already familiar with this bug from a previous mailing list discussion: <a href="http://lists.xenproject.org/archives/html/mirageos-devel/2014-07/msg00131.html">wireshark capture of failed download from mirage-www on ARM</a>.
That turned out be <a href="http://lists.xenproject.org/archives/html/xen-users/2014-07/msg00067.html">a Linux bug</a> - the privileged dom0 code responsible for sending our virtual network packets to the real network becomes confused if two packets occupy the same physical page in memory.</p>

<p>Here’s what happens:</p>

<ol>
  <li>We read 1 MB of data from the disk and send it to the HTTP layer as the next chunk.</li>
  <li>
<a href="https://github.com/mirage/ocaml-cohttp/blob/86394acdb580257ee78cb6976966662575b01bb0/cohttp/transfer_io.ml#L48"><code>Chunked.write</code></a> does the HTTP chunking and sends it to the TCP/IP channel.</li>
  <li>
<a href="https://github.com/mirage/mirage-tcpip/blob/dedd5e0626a3fc8d679e60e92e5fc327c37759bd/channel/channel.ml#L181"><code>Channel.write_string</code></a> writes the HTTP output into pages (aligned 4K blocks of memory).</li>
  <li>
<a href="https://github.com/mirage/mirage-tcpip/blob/dedd5e0626a3fc8d679e60e92e5fc327c37759bd/tcp/pcb.ml#L476"><code>Pcb.writefn</code></a> then determines that each page is too big for a TCP packet and splits each one into smaller chunks, sharing the single underlying page:</li>
</ol>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
</pre></td>
<td><pre><code><span>  <span>let</span> <span>rec</span> <span>writefn</span> <span>pcb</span> <span>wfn</span> <span>data</span> <span>=</span>
</span><span>    <span>let</span> <span>len</span> <span>=</span> <span>Cstruct</span><span>.</span><span>len</span> <span>data</span> <span>in</span>
</span><span>    <span>match</span> <span>write_available</span> <span>pcb</span> <span>with</span>
</span><span>    <span>|</span> <span>0</span> <span>-&gt;</span>
</span><span>      <span>write_wait_for</span> <span>pcb</span> <span>1</span> <span>&gt;&gt;</span>
</span><span>      <span>writefn</span> <span>pcb</span> <span>wfn</span> <span>data</span>
</span><span>    <span>|</span> <span>av_len</span> <span>when</span> <span>av_len</span> <span>&lt;</span> <span>len</span> <span>-&gt;</span>
</span><span>      <span>let</span> <span>first_bit</span> <span>=</span> <span>Cstruct</span><span>.</span><span>sub</span> <span>data</span> <span>0</span> <span>av_len</span> <span>in</span>
</span><span>      <span>let</span> <span>remaing_bit</span> <span>=</span> <span>Cstruct</span><span>.</span><span>sub</span> <span>data</span> <span>av_len</span> <span>(</span><span>len</span> <span>-</span> <span>av_len</span><span>)</span> <span>in</span>
</span><span>      <span>writefn</span> <span>pcb</span> <span>wfn</span> <span>first_bit</span>  <span>&gt;&gt;</span>
</span><span>      <span>writefn</span> <span>pcb</span> <span>wfn</span> <span>remaing_bit</span>
</span><span>    <span>|</span> <span>av_len</span> <span>-&gt;</span>
</span><span>      <span>wfn</span> <span>[</span><span>data</span><span>]</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>My original fix changed <code>mirage-net-xen</code> to wait until the first buffer had been read before sending the second one.
That fixed the retransmissions, but all the waiting meant I still only got 56 KB/s.
Instead, I changed <code>writefn</code> to copy <code>remaining_bit</code> into a new IO page, and with that I got 495 KB/s.</p>

<p>Replacing the filesystem read with a simple <code>String.create</code> of the same length, I got 3.9 MB/s, showing that once again the
FAT filesystem was now the limiting factor.</p>

<h3 id="adding-a-block-cache">Adding a block cache</h3>

<p>I tried adding a <a href="https://github.com/0install/0repo-queue/blob/fat-caching/block_cache.ml">block cache</a> layer between <code>mirage-block-xen</code> and <code>fat-filesystem</code>, like this:</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
<span>18</span>
<span>19</span>
<span>20</span>
<span>21</span>
</pre></td>
<td><pre><code><span><span>module</span> <span>Main</span> <span>(</span><span>C</span> <span>:</span> <span>V1_LWT</span><span>.</span><span>CONSOLE</span><span>)</span>
</span><span>            <span>(</span><span>B</span> <span>:</span> <span>V1_LWT</span><span>.</span><span>BLOCK</span><span>)</span>
</span><span>	    <span>(</span><span>H</span> <span>:</span> <span>Cohttp_lwt</span><span>.</span><span>Server</span><span>)</span> <span>=</span> <span>struct</span>
</span><span>  <span>module</span> <span>BC</span> <span>=</span> <span>Block_cache</span><span>.</span><span>Make</span><span>(</span><span>B</span><span>)</span>
</span><span>  <span>module</span> <span>F</span> <span>=</span> <span>Fat</span><span>.</span><span>Fs</span><span>.</span><span>Make</span><span>(</span><span>BC</span><span>)(</span><span>Io_page</span><span>)</span>
</span><span>  <span>module</span> <span>Q</span> <span>=</span> <span>Upload_queue</span><span>.</span><span>Make</span><span>(</span><span>F</span><span>)</span>
</span><span>
</span><span>  <span>let</span> <span>mem_cache_size</span> <span>=</span> <span>1024</span> <span>*</span> <span>1024</span>	<span>(* 1 MB *)</span>
</span><span>
</span><span>  <span>let</span> <span>start</span> <span>c</span> <span>b</span> <span>http</span> <span>=</span>
</span><span>    <span>Log</span><span>.</span><span>write</span> <span>:=</span> <span>C</span><span>.</span><span>log_s</span> <span>c</span><span>;</span>
</span><span>    <span>Log</span><span>.</span><span>info</span> <span>"start in queue service"</span> <span>&gt;&gt;=</span> <span>fun</span> <span>()</span> <span>-&gt;</span>
</span><span>
</span><span>    <span>BC</span><span>.</span><span>connect</span> <span>(</span><span>b</span><span>,</span> <span>mem_cache_size</span><span>)</span> <span>&gt;&gt;=</span> <span>function</span>
</span><span>    <span>|</span> <span>`</span><span>Error</span> <span>_</span> <span>-&gt;</span> <span>failwith</span> <span>"BC.connect"</span>
</span><span>    <span>|</span> <span>`</span><span>Ok</span> <span>bc</span> <span>-&gt;</span>
</span><span>    <span>F</span><span>.</span><span>connect</span> <span>bc</span> <span>&gt;&gt;=</span> <span>function</span>
</span><span>    <span>|</span> <span>`</span><span>Error</span> <span>_</span> <span>-&gt;</span> <span>failwith</span> <span>"F.connect"</span>
</span><span>    <span>|</span> <span>`</span><span>Ok</span> <span>fs</span> <span>-&gt;</span>
</span><span>    <span>Q</span><span>.</span><span>create</span> <span>fs</span> <span>&gt;&gt;=</span> <span>fun</span> <span>q</span> <span>-&gt;</span>
</span><span><span>...</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>With this in place, upload speed remains at 76 KB/s, but the download speed increases to 1 MB/s (for a 20 MB file, which therefore doesn’t fit in the cache).
This suggests that the FAT filesystem is reading the same disk sectors many times.
Enlarging the memory cache to cover the whole file, the download speed only increases to 1.3 MB/s,
so the FAT code must be doing some inefficient calculations too.</p>

<h3 id="replacing-fat">Replacing FAT</h3>

<p>Since most of my problems seemed to be coming from using FAT, I decided to try a new approach.
I removed all the FAT code and the block cache and changed <code>upload_queue.ml</code> to write directly to the block
device.
With that (no caching), I get:</p>

<table>
  <tbody>
    <tr>
      <td>Upload speed</td>
      <td>2.27 MB/s</td>
    </tr>
    <tr>
      <td>Download speed</td>
      <td>2.46 MB/s</td>
    </tr>
  </tbody>
</table>

<p>That’s not too bad. It’s faster than my Internet connection, which means that the unikernel is no longer the limiting factor.</p>

<p>Here’s the new version: <a href="https://github.com/0install/0repo-queue/blob/master/upload_queue.ml"><code>upload_queue.ml</code></a>.
The big simplification comes from knowing that the queue will spend most of its time empty (another good reason to use a small VM for it).
The code has a <code>next_free_sector</code> which it advances every time an upload starts.
When the queue becomes empty and there are no uploads in progress this variable is reset back to sector 1 (sector 0 holds the index).
This does mean that we may report disk full errors to uploaders even when there is free space on the disk, but this won’t happen in typical usage because the repository downloads things as soon as they’re uploaded (if it does happen, it just means uploaders have to wait a couple of minutes until the repository empties the queue).</p>

<p>Managing the block device manually brought a few more advantages over FAT:</p>

<ol>
  <li>No need to generate random file names for the uploads.</li>
  <li>No need to delete incomplete uploads (we only write the file’s index entry to disk on success).</li>
  <li>The system should recover automatically from filesystem corruption because invalid entries can be detected reliably at boot time and discarded.</li>
  <li>Disk full errors are reported correctly.</li>
  <li>The queue ordering isn’t lost on reboot.</li>
</ol>

<h2 id="conclusions">Conclusions</h2>

<p>Modern operating systems are often extremely complex, but much of this is historical baggage which isn’t needed on a modern system where you’re running a single application as a VM under a hypervisor. Mirage allows you to create very small VMs which contain almost no C code. These VMs should be easier to write, more reliable and more secure.</p>

<p>Creating a bootable OCaml kernel is surprisingly easy, and from there adding support for extra devices is just a matter of pulling in the appropriate libraries. By programming against generic interfaces, you can create code that runs under Linux/Unix/OS X or as a virtual machine under Xen, and switch between configurations using the <code>mirage</code> tool.</p>

<p>Mirage is still very young, and I found many rough edges while writing my queuing service for 0install:</p>

<ul>
  <li>While Linux provides fast, reliable filesystems as standard, Mirage currently only provides a basic FAT implementation.</li>
  <li>Linux provides caching as standard, while you have to implement this yourself on Mirage.</li>
  <li>Error reporting should be a big improvement over C’s error codes, but getting friendly error messages from Mirage is currently difficult.</li>
  <li>The system has clearly been designed for high performance (the APIs generally write to user-provided buffers to avoid copying, much like C libraries do), but many areas have not yet been optimised.</li>
  <li>Buffers often have extra requirements (e.g. must be page-aligned, a single page, immutable, etc) which are not currently captured in the type system, and this can lead to run-time errors which would ideally have been detected at compile time.</li>
</ul>

<p>However, there is <a href="http://openmirage.org/blog/announcing-mirage-20-release">a huge amount of work</a> happening on Mirage right now and it looks like all of these problems are being worked on.
If you’re interested in low-level OS programming and don’t want to mess about with C, Mirage is a lot of fun, and it can be useful for practical tasks already with a bit of effort.</p>

<p>There are still many areas I need to find out more about.
In particular, using the new <a href="http://openmirage.org/blog/introducing-ocaml-tls">pure-OCaml TLS stack</a> to secure the system and trying the <a href="http://openmirage.org/blog/introducing-irmin">Irmin Git-like distributed, branchable storage</a> to provide the queue instead of writing it myself.
I hope to try those soon…</p>
</div>

  </div><br>
<br>
<br>
<br>
<br>
<br>
via roscidus.com http://roscidus.com<br>
<br>
January 4, 2019 at 07:33PM
